### ISSUE TITLE
Inference fails with 4 bit quantization

### ISSUE THREAD

**Issue by gpchowdari** [microsoft/onnxruntime#25631]
### Describe the issue

I am able to export the model to onnx and able to quantize with 8 bit. Infernce works well with 8 bit. But when i try to quantize to 4 bit, inference fails with 4 bit with below error.

INVALID_GRAPH : Load model from sample_convInteger_onnx_4bit.onnx failed:This is an invalid model. Type Error: Type 'tensor(uint4)' of input parameter (conv1.weight_quantized) of operator (ConvInteger) in node (node_Conv_0_quant) is invalid.

Versions:
```
numpy                    2.2.6
onnx                     1.17.0
onnx-ir                  0.1.4
onnxconverter-common     1.14.0
onnxruntime              1.22.0
onnxscript               0.3.2
protobuf                 3.20.2
torch                    2.7.1
```

### To reproduce

```
import torch
import torch.nn as nn
import torch.nn.functional as F

from torch.export.dynamic_shapes import Dim
import onnx
from onnxruntime.quantization import quantize_dynamic, QuantType

import onnxruntime as ort
import time


def run_inference(onnx_model,input):
    opts = ort.SessionOptions()
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") 
    opts.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL

    start_time = time.time()
    ort_session = ort.InferenceSession(onnx_model,opts)
    def to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
    inputs = {
                "x": to_numpy(input.to(device)),}    
    x = ort_session.run(
                    None,
                    inputs
                )
    print("Output shape:", output.shape)
    
def quantize_onnx_model(onnx_model_path, quantized_model_path,quantType):
     
     model = onnx.load(onnx_model_path)
     with torch.no_grad():  
        quantize_dynamic(model,
                    quantized_model_path,
                    per_channel=True,                    
                    weight_type=quantType,
                    use_external_data_format=True,
                    extra_options = {"DefaultTensorType":onnx.TensorProto.FLOAT,},
                )
        
     print("ONNX Quantization Done")         

def onnx_export_s2tt(n_model,onnx_name,input):  

  with torch.no_grad():    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") 
    model = n_model.eval()
    with torch.no_grad(): 
        onnx_program = torch.onnx.export(
        model,
        input,
        input_names=['x'],
        output_names=["output"],
        export_params=True,
        dynamo=True,
        do_constant_folding=True,
        optimize=True,
        verify = True,
        opset_version=22,
        )
        onnx_program.save(onnx_name)
        print('ONNX Conversion Done')

class ConvInteger(nn.Module):
    def __init__(self, num_classes=10):
        super(ConvInteger, self).__init__()
        self.conv1 = nn.Conv2d(in_channels=1, out_channels=16, kernel_size=3, stride=1, padding=1)
        self.conv2 = nn.Conv2d(in_channels=16, out_channels=32, kernel_size=3, stride=1, padding=1)
        self.conv3 = nn.Conv2d(in_channels=32, out_channels=64, kernel_size=3, stride=1, padding=1)
        
        self.global_avg_pool = nn.AdaptiveAvgPool2d((1, 1))
        
        self.fc = nn.Linear(64, num_classes)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = F.relu(self.conv3(x))
        
        x = self.global_avg_pool(x)
        
        x = torch.flatten(x, 1)
        
        x = self.fc(x)
        return x

if __name__ == "__main__":
    model = ConvInteger(num_classes=10)
    print(model)
    
    input_tensor = torch.randn(1, 1, 128, 128)  
    output = model(input_tensor)
    print("Output shape:", output.shape)
    onnx_name = 'sample_convInteger_onnx.onnx'
    onnx_q4_name = 'sample_convInteger_onnx_4bit.onnx'
    onnx_export_s2tt(model,onnx_name,input_tensor)
    run_inference(onnx_name,input_tensor)
    quantize_onnx_model(onnx_name,onnx_q4_name,QuantType.QUInt4)
    run_inference(onnx_q4_name,input_tensor)
```

### Platform

Linux

### OS Version

22.0.4

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.22.0

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

**Comment by DakeQQ** [comment]
Currently, it recommend using `matmul_nbits_quantizer()` instead of `quantize_dynamic()`. Please refer to the example below demonstrating its usage. Note: For 8-bit quantization + `CPUExecutionProvider`, the `quantize_dynamic()` still outperforms `matmul_nbits_quantizer()`.


```python
import onnx
from pathlib import Path
from onnxruntime.transformers.optimizer import optimize_model
from onnxruntime.quantization import (
    matmul_nbits_quantizer,  # onnxruntime >= 1.22.0
    quant_utils
)

# Path Setting
original_folder_path = r"C:\Users\Downloads\Model_ONNX"                          # The original folder.
quanted_folder_path = r"C:\Users\Downloads\Model_ONNX_Optimized"                 # The optimized folder.
model_path = os.path.join(original_folder_path, "Model.onnx")                    # The original fp32 model path.
quanted_model_path = os.path.join(quanted_folder_path, "Model_Optimized.onnx")   # The optimized model stored path.


algorithm = "DEFAULT"                                                            # ["DEFAULT", "RTN", "HQQ",], HQQ will very slow both in quant and inference.
bits = 4                                                                         # [4, 8]
op_types = ["MatMul"]                                                            # ["MatMul", "Gather"]; Adding Gather may get errors.
quant_axes = [0]                                                                 # Target axes to quant the quant data.
block_size = 128                                                                 # [32, 64, 128, 256]; A smaller block_size yields greater accuracy but increases quantization time and model size.
accuracy_level = 4                                                               # 0:default, 1:fp32, 2:fp16, 3:bf16, 4:int8
quant_symmetric = False                                                          # False may get more accuracy.
nodes_to_exclude = None                                                          # Set the node names here. Such as: ["/layers.0/mlp/down_proj/MatMul"]


# Start Weight-Only Quantize
model = quant_utils.load_model_with_shape_infer(Path(model_path))

if algorithm == "RTN":
    quant_config = matmul_nbits_quantizer.RTNWeightOnlyQuantConfig(
        quant_format=quant_utils.QuantFormat.QOperator,
        op_types_to_quantize=tuple(op_types)
    )
elif algorithm == "HQQ":
    quant_config = matmul_nbits_quantizer.HQQWeightOnlyQuantConfig(
        bits=bits,
        block_size=block_size,
        axis=quant_axes[0],
        quant_format=quant_utils.QuantFormat.QOperator,
        op_types_to_quantize=tuple(op_types),
        quant_axes=tuple((op_types[i], quant_axes[i]) for i in range(len(op_types)))
    )
else:
    quant_config = matmul_nbits_quantizer.DefaultWeightOnlyQuantConfig(
        block_size=block_size,
        is_symmetric=quant_symmetric,
        accuracy_level=accuracy_level,
        quant_format=quant_utils.QuantFormat.QOperator,
        op_types_to_quantize=tuple(op_types),
        quant_axes=tuple((op_types[i], quant_axes[i]) for i in range(len(op_types)))
    )
quant_config.bits = bits
quant = matmul_nbits_quantizer.MatMulNBitsQuantizer(
    model,
    block_size=block_size,
    is_symmetric=quant_symmetric,
    accuracy_level=accuracy_level,
    quant_format=quant_utils.QuantFormat.QOperator,
    op_types_to_quantize=tuple(op_types),
    quant_axes=tuple((op_types[i], quant_axes[i]) for i in range(len(op_types))),
    algo_config=quant_config,
    nodes_to_exclude=nodes_to_exclude
)
quant.process()
quant.model.save_model_to_file(
    quanted_model_path,
    True                                         # save_as_external_data
)
```

**Comment by gpchowdari** [comment]
@DakeQQ Thanks for the pointers. But when i checked with the suggested approach, The model size is not reduced much. Actual onnx model size is ~8gb. with 8 bit quantization using `quantize_dynamic()` my model size comes around 1.8GB. With 4bit it comes around ~1GB ( Though inference fails). With `matmul_nbits_quantizer()` with 4 bit, My model size is around ~6GB.

**Comment by DakeQQ** [comment]
The `matmul_nbits_quantizer()` has two key limitations: it only works on 2D constant weight data and does not support Embedding weight quantization. When using this function, ensure you correctly set the quantization axis for your model's constant weight data. For example, with a weight shape of [1024, 10], setting the quantization axis to 1 will lead to inefficient quantization. 

Using a 1.5B LLM as an example, `matmul_nbits_quantizer()` is expected to reduce a 6.5GB FP32 model to approximately 1.3GB in Int4 format.
